### ISSUE TITLE
BT_REGS_NOT_FOUND when bt a disk dump file

### ISSUE THREAD

**Issue by haoren3696** [crash-utility/crash#5]
Hi, I got result below when using bt command。

crash> bt
PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
bt: WARNING: cannot determine starting stack frame for task ffffffd095bde

the crash version is 7.1.5 and the vmcore is produced by a kernel panic injection. I wrote a panic in a module init function and insmod this module to make system panic.

It seems like some thing wrong with arm64_get_dumpfile_stackframe in arm64. When the panic happend on the kernel stack of  a user space task.

        if (user_mode(ptregs)) {
                frame->sp = user_stack_pointer(ptregs);
                frame->fp = user_frame_pointer(ptregs);
                if (is_kernel_text(frame->pc) ||
                    !in_user_stack(bt->tc->task, frame->sp)) {
                        error(WARNING,
                            "corrupt prstatus? pstate=0x%lx, but no user frame found\n",
                                ptregs->pstate);
                        bt->flags |= BT_REGS_NOT_FOUND;
                        return FALSE;
                }
                bt->flags |= BT_USER_SPACE;
        } else {
                frame->sp = ptregs->sp;
                frame->fp = ptregs->regs[29];
        }

I replace this part with implementation in crash-7.1.3 below ,and got the right result.

        frame->pc = ptregs->pc;
        frame->fp = ptregs->regs[29];

        if (!is_kernel_text(frame->pc) &&
            in_user_stack(bt->tc->task, frame->sp))
                bt->flags |= BT_USER_SPACE;

        if (arm64_in_kdump_text(bt, frame))
                bt->flags |= BT_KDUMP_ADJUST;

**Comment by crash-utility** [comment]
----- Original Message -----
> Hi, I got result below when using bt command。
> 
> crash> bt
> PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
> bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
> bt: WARNING: cannot determine starting stack frame for task ffffffd095bde
> 
> the crash version is 7.1.5 and the vmcore is produced by a kernel panic
> injection. I wrote a panic in a module init function and insmod this module
> to make system panic.
> 
> It seems like some thing wrong with arm64_get_dumpfile_stackframe in arm64.
> When the panic happend on the kernel stack of  a user space task.
> 
>         if (user_mode(ptregs)) {
>                 frame->sp = user_stack_pointer(ptregs);
>                 frame->fp = user_frame_pointer(ptregs);
>                 printf("xzpeng:sp=%lx fp=%lx pc=%lx
>                 ptregs->pstate:%lx\n",frame->sp,frame->fp,frame->pc,ptregs->pstate);
>                 if (is_kernel_text(frame->pc) ||
>                     !in_user_stack(bt->tc->task, frame->sp)) {
>                         error(WARNING,
>                             "corrupt prstatus? pstate=0x%lx, but no user
>                             frame found\n",
>                                 ptregs->pstate);
>                         bt->flags |= BT_REGS_NOT_FOUND;
>                         return FALSE;
>                 }
>                 bt->flags |= BT_USER_SPACE;
>         } else {
>                 frame->sp = ptregs->sp;
>                 frame->fp = ptregs->regs[29];
>         }
> 

What is the output of your "printf("xzpeng: ..." statement?

Dave

**Comment by haoren3696** [comment]
> What is the output of your "printf("xzpeng: ..." statement?

crash> bt
xzpeng:sp=ffffffc04c98fa80 fp=ffffffc04c98fa80 pc=ffffffc0002aab2c ptregs->pstate:20000000
PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
bt: WARNING: cannot determine starting stack frame for task ffffffd095bde400

**Comment by haoren3696** [comment]
here is the result i get from crash-7.1.3

```
crash> bt
PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
 #0 [ffffffc04c98fa20] crash_kexec at ffffffc0002aaaac
 #1 [ffffffc04c98fa90] panic at ffffffc000690ff0
 #2 [ffffffc04c98fc10] lkdtm_do_action at ffffffbfa2ccd434 [lkdtm]
 #3 [ffffffc04c98fce0] direct_entry at ffffffbfa2ccdc94 [lkdtm]
 #4 [ffffffc04c98fd70] __vfs_write at ffffffc00037eaac
 #5 [ffffffc04c98fda0] vfs_write at ffffffc00037f4fc
 #6 [ffffffc04c98fe40] sys_write at ffffffc000380158
 #7 [ffffffc04c98fe80] __sys_trace_return at ffffffc000203e8c
     PC: 0000000000000000   LR: 0000007f8611a5d8   SP: 0000000020000000
    X29: 0000007fc9b1aa40  X28: 0000007f860bbb48  X27: 0000007fc9b1aa40
    X26: 000000000051a360  X25: 0000000000518580  X24: 0000000000000001
    X23: 0000000000508000  X22: 00000000004bff78  X21: 000000000050b000
    X20: 0000000000000006  X19: 0000007f861c84c0  X18: 0000007f86000000
    X17: 0000000000000006  X16: 0000000000000000  X15: 0000007f860b92c0
    X14: 0000000000000000  X13: 002792ff16a58cf2  X12: 0000000000000000
    X11: 00000003e8000000  X10: 0000000000000018   X9: 0000007fc9b1aaa0
     X8: 00000000ffffffd0   X7: 8101010101010100   X6: 0000000000000040
     X5: a4242424242454ff   X4: 0000000000518580   X3: 0000007f860922ec
     X2: 0000000000000000   X1: 0000000000000000   X0: 0000000000000006
    ORIG_X0: 0000000000000001  SYSCALLNO: 40  PSTATE: 00000000

```

**Comment by crash-utility** [comment]
----- Original Message -----
> 
> > What is the output of your "printf("xzpeng: ..." statement?
> 
> crash> bt
> xzpeng:sp=ffffffc04c98fa80 fp=ffffffc04c98fa80 pc=ffffffc0002aab2c
> ptregs->pstate:20000000
> PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
> bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
> bt: WARNING: cannot determine starting stack frame for task ffffffd095bde400
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/crash-utility/crash/issues/5#issuecomment-221443349

The 7.1.3 code did not check the pstate, which indicates that the task is in
user-space mode.  I think this is a kdump bug, but perhaps a workaround could
be added where the pstate warning message could be displayed, but if the sp and
fp are in the current stack and the pc is kernel text, it could still utilize
them.  BTW, is your kernel running on bare metal or is it a KVM guest?

Anyway, try this hack and let me know how it works:

diff --git a/arm64.c b/arm64.c
index 34c8c59..505dcb7 100644
--- a/arm64.c
+++ b/arm64.c
@@ -1711,11 +1711,14 @@ arm64_get_dumpfile_stackframe(struct bt_info *bt, struct arm64_stackframe *frame
                        error(WARNING, 
                            "corrupt prstatus? pstate=0x%lx, but no user frame found\n",
                                ptregs->pstate);
+                       if (INSTACK(frame->sp, bt) && INSTACK(frame->fp, bt))
+                               goto try_kernel;
                        bt->flags |= BT_REGS_NOT_FOUND;
                        return FALSE;
                }
                bt->flags |= BT_USER_SPACE;
        } else {
+try_kernel:
                frame->sp = ptregs->sp;
                frame->fp = ptregs->regs[29];
        }

Thanks,
  Dave

**Comment by haoren3696** [comment]
> The 7.1.3 code did not check the pstate, which indicates that the task is in
> user-space mode.  I think this is a kdump bug, but perhaps a workaround could
> be added where the pstate warning message could be displayed, but if the sp and
> fp are in the current stack and the pc is kernel text, it could still utilize
> them.  BTW, is your kernel running on bare metal or is it a KVM guest?

my kernel is running on bare metal

> Anyway, try this hack and let me know how it works:
> 
> ```
> diff --git a/arm64.c b/arm64.c
> index 34c8c59..505dcb7 100644
> --- a/arm64.c
> +++ b/arm64.c
> @@ -1711,11 +1711,14 @@ arm64_get_dumpfile_stackframe(struct bt_info *bt, struct arm64_stackframe *frame
>                         error(WARNING,
>                             "corrupt prstatus? pstate=0x%lx, but no user frame found\n",
>                                 ptregs->pstate);
> +                       if (INSTACK(frame->sp, bt) && INSTACK(frame->fp, bt))
> +                               goto try_kernel;
>                         bt->flags |= BT_REGS_NOT_FOUND;
>                         return FALSE;
>                 }
>                 bt->flags |= BT_USER_SPACE;
>         } else {
> +try_kernel:
>                 frame->sp = ptregs->sp;
>                 frame->fp = ptregs->regs[29];
>         }
> ```

yes,this patch fix my problem.

```
crash> bt
PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
 #0 [ffffffc04c98fa20] crash_kexec at ffffffc0002aaaac
 #1 [ffffffc04c98fa90] panic at ffffffc000690ff0
 #2 [ffffffc04c98fc10] lkdtm_do_action at ffffffbfa2ccd434 [lkdtm]
 #3 [ffffffc04c98fce0] direct_entry at ffffffbfa2ccdc94 [lkdtm]
 #4 [ffffffc04c98fd70] __vfs_write at ffffffc00037eaac
 #5 [ffffffc04c98fda0] vfs_write at ffffffc00037f4fc
 #6 [ffffffc04c98fe40] sys_write at ffffffc000380158
 #7 [ffffffc04c98fe80] __sys_trace_return at ffffffc000203e8c
     PC: 0000000000000000   LR: 0000007f8611a5d8   SP: 0000000020000000
    X29: 0000007fc9b1aa40  X28: 0000007f860bbb48  X27: 0000007fc9b1aa40
    X26: 000000000051a360  X25: 0000000000518580  X24: 0000000000000001
    X23: 0000000000508000  X22: 00000000004bff78  X21: 000000000050b000
    X20: 0000000000000006  X19: 0000007f861c84c0  X18: 0000007f86000000
    X17: 0000000000000006  X16: 0000000000000000  X15: 0000007f860b92c0
    X14: 0000000000000000  X13: 002792ff16a58cf2  X12: 0000000000000000
    X11: 00000003e8000000  X10: 0000000000000018   X9: 0000007fc9b1aaa0
     X8: 00000000ffffffd0   X7: 8101010101010100   X6: 0000000000000040
     X5: a4242424242454ff   X4: 0000000000518580   X3: 0000007f860922ec
     X2: 0000000000000000   X1: 0000000000000000   X0: 0000000000000006
    ORIG_X0: 0000000000000001  SYSCALLNO: 40  PSTATE: 00000000

```

**Comment by crash-utility** [comment]
----- Original Message -----
> > The 7.1.3 code did not check the pstate, which indicates that the task is in
> > user-space mode.  I think this is a kdump bug, but perhaps a workaround could
> > be added where the pstate warning message could be displayed, but if the sp and
> > fp are in the current stack and the pc is kernel text, it could still utilize
> > them.  BTW, is your kernel running on bare metal or is it a KVM guest?
> 
> my kernel is running on bare metal
> 
> > Anyway, try this hack and let me know how it works:
> > 
> > ```
> > diff --git a/arm64.c b/arm64.c
> > index 34c8c59..505dcb7 100644
> > --- a/arm64.c
> > +++ b/arm64.c
> > @@ -1711,11 +1711,14 @@ arm64_get_dumpfile_stackframe(struct bt_info *bt,
> > struct arm64_stackframe *frame
> >                         error(WARNING,
> >                             "corrupt prstatus? pstate=0x%lx, but no user
> >                             frame found\n",
> >                                 ptregs->pstate);
> > +                       if (INSTACK(frame->sp, bt) && INSTACK(frame->fp, bt))
> > +                               goto try_kernel;
> >                         bt->flags |= BT_REGS_NOT_FOUND;
> >                         return FALSE;
> >                 }
> >                 bt->flags |= BT_USER_SPACE;
> >         } else {
> > +try_kernel:
> >                 frame->sp = ptregs->sp;
> >                 frame->fp = ptregs->regs[29];
> >         }
> > ```
> 
> yes,this patch fix my problem.
> 
> ```
> crash> bt
> PID: 25723  TASK: ffffffd095bde400  CPU: 0   COMMAND: "bash"
> bt: WARNING: corrupt prstatus? pstate=0x20000000, but no user frame found
>  #0 [ffffffc04c98fa20] crash_kexec at ffffffc0002aaaac
>  #1 [ffffffc04c98fa90] panic at ffffffc000690ff0
>  #2 [ffffffc04c98fc10] lkdtm_do_action at ffffffbfa2ccd434 [lkdtm]
>  #3 [ffffffc04c98fce0] direct_entry at ffffffbfa2ccdc94 [lkdtm]
>  #4 [ffffffc04c98fd70] __vfs_write at ffffffc00037eaac
>  #5 [ffffffc04c98fda0] vfs_write at ffffffc00037f4fc
>  #6 [ffffffc04c98fe40] sys_write at ffffffc000380158
>  #7 [ffffffc04c98fe80] __sys_trace_return at ffffffc000203e8c
>      PC: 0000000000000000   LR: 0000007f8611a5d8   SP: 0000000020000000
>     X29: 0000007fc9b1aa40  X28: 0000007f860bbb48  X27: 0000007fc9b1aa40
>     X26: 000000000051a360  X25: 0000000000518580  X24: 0000000000000001
>     X23: 0000000000508000  X22: 00000000004bff78  X21: 000000000050b000
>     X20: 0000000000000006  X19: 0000007f861c84c0  X18: 0000007f86000000
>     X17: 0000000000000006  X16: 0000000000000000  X15: 0000007f860b92c0
>     X14: 0000000000000000  X13: 002792ff16a58cf2  X12: 0000000000000000
>     X11: 00000003e8000000  X10: 0000000000000018   X9: 0000007fc9b1aaa0
>      X8: 00000000ffffffd0   X7: 8101010101010100   X6: 0000000000000040
>      X5: a4242424242454ff   X4: 0000000000518580   X3: 0000007f860922ec
>      X2: 0000000000000000   X1: 0000000000000000   X0: 0000000000000006
>     ORIG_X0: 0000000000000001  SYSCALLNO: 40  PSTATE: 00000000

Ok, thanks.  

A simultaneous discussion is also occurring on the crash-utility mailing list.
It is essentially a kernel/kdump bug, but I will add this patch anyway for
backwards-compatibility, and because it's a regression in the crash utility:

  Re: [Crash-utility] [PATCH v2] arm64: fix kernel memory map handling for kaslr-enabled kernel
  https://www.redhat.com/archives/crash-utility/2016-May/msg00062.html

I prefer to not use this github "issues" facility for discussing crash bugs.
I invite you to join the crash-utility mailing list, which is the ideal place
to post bugs/patches/questions:

  https://www.redhat.com/mailman/listinfo/crash-utility

Thanks,
  Dave

**Comment by crash-utility** [comment]
Fixed in commit bced530fcf21f803e2f758ef0279d07d70fca220
